tests/unit-tests/dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -305,23 +443,34 @@ void verify_int32_union(DDS::DynamicType_var dt, const DataView& expected_cdr)
 {
   XTypes::DynamicDataImpl data(dt);
   DDS::ReturnCode_t ret = data.set_int32_value(XTypes::DISCRIMINATOR_ID, E_INT32);
-  EXPECT_EQ(ret, DDS::RETCODE_OK);
+  EXPECT_EQ(ret, DDS::RETCODE_ERROR);
```

Comment:
I assume that either the test was wrong before or the new implementation breaks it.  Can you specify which one and why?

Reply:
Yes, the new implementation breaks it. Before we allowed setting the discriminator directly and that can leave the container of a union in a state where it has the discriminator but not the member it selects. I updated how we handle union to follow the IDL-to-C++ mapping more strictly, which doesn't let it set the discriminator in this case. The state of the container for a union now maintains an invariant that it either is empty or contains valid data for the union, meaning it has the discriminator and an activated member if that discriminator selects it. I think it'll be good to update the DevGuide for the behavior of union in DynamicData.

Reply:
Is the empty state visible to the user?  If the user creates a dynamic union and then writes the sample is it an error or does it write some valid value?  (This is one place where the "C++ classic" and C++11 mappings differ, so there's no single right way to do it.)

Reply:
It'll write a valid value constructed from the default value for the discriminator and the default value for a selected member if one is selected.

The empty state is probably not visible to the user since we're returning a default value when the user tries to get a member from an empty union.